### ISSUE TITLE
"Cannot find module 'lodash'" and "Cannot find module '../lang/deepClone'"

### ISSUE THREAD

**Issue by guillehorno** [yeoman/generator-webapp_DEPRECATED#349]
I started having the same problem as mentioned in issue #348 

I never saw a test folder on the root of my webapp and I suspected that the problem could be related with that. I tried to compare the folder structure generating a fresh webapp with yeoman and started having different issues.

During the `npm install` phase of the generation, this meessages show up.

```
> pngquant-bin@0.1.7 postinstall /Users/malbecmedia/Desarrollo/tmp/webapp5/node_modules/grunt-contrib-imagemin/node_modules/image-min/node_modules/pngquant-bin
> node index.js


module.js:340
    throw err;
          ^
Error: Cannot find module '../lang/deepClone'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/Users/malbecmedia/Desarrollo/tmp/webapp5/node_modules/grunt-contrib-imagemin/node_modules/image-min/node_modules/mout/object/merge.js:2:17)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
```

And at the end of the process, this one:

```
grunt-contrib-imagemin@0.6.1 node_modules/grunt-contrib-imagemin
├── pretty-bytes@0.1.0
├── mkdirp@0.3.5
├── chalk@0.4.0 (has-color@0.1.7, ansi-styles@1.0.0, strip-ansi@0.1.1)
├── async@0.2.10
└── image-min@0.2.3 (map-key@0.1.1, win-spawn@2.0.0, filesize@2.0.3, stream-combiner@0.0.4, multipipe@0.0.2, through2@0.4.1, concat-stream@1.4.5, mout@0.9.1, gifsicle@0.1.5, jpegtran-bin@0.2.6, optipng-bin@0.3.3)

module.js:340
    throw err;
    ^
Error: Cannot find module 'lodash'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/usr/local/lib/node_modules/generator-mocha/node_modules/yeoman-generator/lib/util/log.js:3:9)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
```

If I do a  `npm cache clean` and try to generate a new webapp, the `'../lang/deepClone'` error disappears (only after a `npm cache clean`) but the lodash error always shows up.

I also noticed that the `bower_components` folder shows up inside the `app` folder again, making the `grunt serve` task throw this error:

```
Warning: Syntax error: File to import not found or unreadable: ../../bower_components/bootstrap-sass-official/vendor/assets/stylesheets/bootstrap.scss.
```

The bower_components change its strange, specially when it was moved out of the app folder on #171 

I installed latest versions of Yeoman and the generator. 

Thanks!

**Comment by guillehorno** [comment]
I'm sorry, I referenced the wrong issue at the end of my message, I just fixed it. Thanks!

**Comment by FunTimeCoding** [comment]
I get the error involving pngquant-bin@0.1.7 too.

**Comment by eddiemonge** [comment]
@guillehorno What version of npm, node, yo and webapp are you using? Whats your output of `npm config list`?

**Comment by guillehorno** [comment]
node: v0.10.26
npm: 1.3.25
yeoman: 1.1.2
generator-webapp@0.4.9

`npm config list` outputs this:

```
; cli configs
registry = "https://registry.npmjs.org/"

; node bin location = /usr/local/bin/node
; cwd = /Users/malbecmedia/Desarrollo/proyectos/TeamPuntoPagos/puntopagos-website
; HOME = /Users/malbecmedia
; 'npm config ls -l' to show all defaults.
```

**Comment by eddiemonge** [comment]
Can you try the following:

```
git clone git@github.com:yeoman/generator-webapp.git
cd generator-webapp
npm install -g generator-mocha
npm link
mkdir ~/Desktop/webappTest && cd $_
yo webapp
```

And see if that fixes it for you?

**Comment by guillehorno** [comment]
I updated to the latest versions of yeoman and generator-webapp and my original issues seem to be fixed. Now with a brand new generation of a webapp using the generator everything works fine, except some weird path issues.

Given that the bower_components folder lives inside the app one, this line in the main.scss has an extra `../` (it gives an error during `grunt build`)

```
// bower:scss
@import '../../bower_components/bootstrap-sass-official/vendor/assets/stylesheets/bootstrap.scss';
// endbower
```

Should be

```

// bower:scss
@import '../bower_components/bootstrap-sass-official/vendor/assets/stylesheets/bootstrap.scss';
// endbower
```

Also, it doesn't seem to find modernizr (I assume due to the location of the bower_componets folder). I had to change the modernizr section on Gruntfile.js to:

```
modernizr: {
            dist: {
                devFile: '<%= config.app %>/bower_components/modernizr/modernizr.js',
                outputFile: '<%= config.dist %>/scripts/vendor/modernizr.js',
                files: {
                    src: [
                        '<%= config.dist %>/scripts/{,*/}*.js',
                        '<%= config.dist %>/styles/{,*/}*.css',
                        '!<%= config.dist %>/scripts/vendor/*'
                    ]
                },
                uglify: true
            }
        },
```

I added the `<%= config.app %>` preceding the location of the bower_components folder on the `devFile` line

If you want I can move this to a new issue so you can close the original one.

Thanks!

**Comment by eddiemonge** [comment]
thats a known issue that i have yet to get to in order to fix

**Comment by eddiemonge** [comment]
This is a bug in `grunt-contrib-imagemin`. Please open an issue there for them to troubleshoot.

**Comment by pvfhv** [comment]
全局安装lodash包，我在yo webapp时提示的是underscore.string缺少，npm install -g underscore.string解决了，同样你可以在全局下安装一下lodash

**Comment by silvenon** [comment]
We don't use lodash anymore, but [underscore.string](https://www.npmjs.com/package/underscore.string), a subset of Underscore. That should be installed globally because #562 isn't released yet.
